pydantic/types.py

```diff
@@ -687,6 +704,102 @@ def get_secret_value(self) -> bytes:
         return self._secret_value
 
 
+class PaymentCardBrand(Enum):
+    amex = 'American Express'
+    mastercard = 'Mastercard'
+    visa = 'Visa'
+    other = 'other'
+
+    def __str__(self) -> str:
+        return self.value
+
+
+class PaymentCardNumber(str):
+    """
+    Based on: https://en.wikipedia.org/wiki/Payment_card_number
+    """
+
+    strip_whitespace: ClassVar[bool] = True
+    min_length: ClassVar[int] = 12
+    max_length: ClassVar[int] = 19
+    bin: str
+    last4: str
+    brand: PaymentCardBrand
+
+    def __init__(self, card_number: str):
+        self.bin = card_number[:6]
+        self.last4 = card_number[-4:]
+        self.brand = self.get_brand(card_number)
+
+    @classmethod
+    def __get_validators__(cls) -> 'CallableGenerator':
+        yield not_none_validator
+        yield str_validator
+        yield constr_strip_whitespace
+        yield constr_length_validator
+        yield cls.validate_digits
+        yield cls.validate_luhn_check_digit
+        yield cls
+        yield cls.validate_length_for_brand
+
+    @property
+    def masked(self) -> str:
+        num_masked = len(self) - 10  # len(bin) + len(last4) == 10
+        return f'{self.bin}{"*" * num_masked}{self.last4}'
+
+    @classmethod
+    def validate_digits(cls, card_number: str) -> str:
+        if not card_number.isdigit():
+            raise errors.NotDigitError
+        return card_number
+
+    @classmethod
+    def validate_luhn_check_digit(cls, card_number: str) -> str:
+        """
+        Based on: https://en.wikipedia.org/wiki/Luhn_algorithm
+        """
+        sum_ = int(card_number[-1])
+        length = len(card_number)
+        parity = length % 2
+        for i in range(length - 1):
+            digit = int(card_number[i])
+            if i % 2 == parity:
+                digit *= 2
+            sum_ += digit
+        valid = sum_ % 10 == 0
+        if not valid:
+            raise errors.LuhnValidationError
+        return card_number
+
+    @classmethod
+    def validate_length_for_brand(cls, card_number: 'PaymentCardNumber') -> 'PaymentCardNumber':
+        """
+        Validate length based on BIN for major brands:
+        https://en.wikipedia.org/wiki/Payment_card_number#Issuer_identification_number_(IIN)
+        """
+        required_lengths = dict(visa=16, mastercard=16, amex=15)
+        try:
+            required_length = required_lengths[card_number.brand.name]
+        except KeyError:
+            pass
+        else:
+            if len(card_number) != required_length:
+                raise errors.InvalidLengthForBrand(brand=card_number.brand, required_length=required_length)
+        return card_number
+
+    @staticmethod
+    def get_brand(card_number: str) -> PaymentCardBrand:
```

Comment:
would it also be useful to have

```py
@property
def brand(self):
    return self.get_brand(self)
```
?

Reply:
That was my preference, but `cython` was not a fan: https://travis-ci.org/samuelcolvin/pydantic/builds/579543856

Reply:
@matin It's usually possible to get around these sorts of cython issues by tricking it. In particular, I think if you annotate the `get_brand` function with `Union[str]` instead of `str`, it should prevent cython from causing a type error. Alternatively, you could manually cast to str: `return self.get_brand(str(self))`

Reply:
Even if that worked, would it be preferable to what we have now?

If we did add those kinds of workarounds, we'd also need to add comments to specify _why_ we're doing it.

Reply:
I believe the `Union[str]` workaround exists in some other places in the code from the initial effort to make it fully cythonized. In general, if the meaning is unambiguous (e.g., `Union[str]` clearly means `str` even if it looks a little weird), and changing it causes CI to break, I'm personally okay with it.

If we wanted to use a property, I think the following would be fine:
```python
@property
def brand(self):
    return self.get_brand(str(self))
```
Despite the potentially confusing cast with str, since changing it would break CI, I think it doesn't really need more documentation (though I also think a `# cast for cython` would be fine).

At any rate, since `self.brand` is already a thing set in the `__init__`, I don't think we need the property, and I think it is better implemented the way it is now since I expect the computation savings are likely to outweigh the memory savings in importance. (When `cached_property` becomes a thing in python 3.8, I think that would be preferable to either of these approaches, but would likely play most nicely with the property-based approach.)

Reply:
In that case, ok to keep it as is?

Reply:
I think so. I would defer to @samuelcolvin though.

Also, sorry I wasn't clear -- my point with these responses was *not* to ask you to change the implementation so much as it was to provide a workaround for the cython issue, since it seemed based on your comment that that might be the only reason it wasn't implemented in a different way. I'm personally fine with this implementation instead though.

Reply:
I'm fine with how it is.

Let's just make this a private method to avoid confusion.